Provide way to dynamically create builtin-operator symbols off of a compilation

CyrusNajmabadi opened the pull request:
Closes https://github.com/dotnet/roslyn/issues/63186

API approved in https://github.com/dotnet/roslyn/issues/63186.

Todo:
- [x] Tests

Youssef1313 commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3821,82 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            Debug.Assert(csharpLeftType is not null);
+            Debug.Assert(csharpRightType is not null);
+
+            var nameOk = name
+                is WellKnownMemberNames.CheckedAdditionOperatorName
+                or WellKnownMemberNames.AdditionOperatorName
+                or WellKnownMemberNames.BitwiseAndOperatorName
+                or WellKnownMemberNames.CheckedDivisionOperatorName
+                or WellKnownMemberNames.DivisionOperatorName
+                or WellKnownMemberNames.EqualityOperatorName
+                or WellKnownMemberNames.GreaterThanOperatorName
+                or WellKnownMemberNames.GreaterThanOrEqualOperatorName
+                or WellKnownMemberNames.LeftShiftOperatorName
+                or WellKnownMemberNames.LessThanOperatorName
+                or WellKnownMemberNames.LessThanOrEqualOperatorName
+                or WellKnownMemberNames.CheckedMultiplyOperatorName
+                or WellKnownMemberNames.MultiplyOperatorName
+                or WellKnownMemberNames.BitwiseOrOperatorName
+                or WellKnownMemberNames.InequalityOperatorName
+                or WellKnownMemberNames.ModulusOperatorName
+                or WellKnownMemberNames.RightShiftOperatorName
+                or WellKnownMemberNames.UnsignedRightShiftOperatorName
+                or WellKnownMemberNames.CheckedSubtractionOperatorName
+                or WellKnownMemberNames.SubtractionOperatorName
+                or WellKnownMemberNames.ExclusiveOrOperatorName;
```
Does this work?

```suggestion
            var nameOk = SyntaxFacts.GetOperatorKind(name) != SyntaxKind.None;
```

Youssef1313 replied:
My suggestion isn't correct as we want to handle binary operators only in this path.

My concern is that we can forget updating this public API if new operators are added in future. It would be great if we can have a Debug assert in `GetOperatorKind` that one of the two APIs handle the new operator.

CyrusNajmabadi replied:
no, that would support a superset of actually supported operator names.

Youssef1313 replied:
@CyrusNajmabadi Yup, already noticed that and commented. 

> My concern is that we can forget updating this public API if new operators are added in future. It would be great if we can have a Debug assert in `GetOperatorKind` that one of the two APIs handle the new operator.

This is still concerning, I think.

CyrusNajmabadi replied:
Do you have a suggestion on what this assert would do?  It's unclear to me that GetOperatorKind will always map to something one of these can create.

Youssef1313 replied:
> Do you have a suggestion on what this assert would do? It's unclear to me that GetOperatorKind will always map to something one of these can create.

Maybe break the two `nameOk` checks into a two helpers (`IsBinaryOperatorName` and `IsUnaryOperatorName`) and for "none `SyntaxKind.None`" returns, we assert that one of those two helpers return true. Not sure if that makes sense

Youssef1313 replied:
> break the two nameOk checks into a two helpers

alternatively, we could have a "fake debug only compilation" that we call the APIs on for the purpose of the assert.

Youssef1313 commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1488,63 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+        /// <param name="isChecked">Whether or not this is a checked binary operator.  For example <c>int int.operator
+        /// checked +(int v1, int v2)</c>.</param>
+        /// <returns></returns>
+        public IMethodSymbol CreateBuiltinOperator(string name, ITypeSymbol returnType, ITypeSymbol leftType, ITypeSymbol rightType, bool isChecked)
```
The signature uses a string parameter, while the referenced issue with approved API uses `BinaryOperatorKind`. Is that intentional?

CyrusNajmabadi replied:
yes.  going down this path i found that that we needed to map the enum to an enum just to get this string.  and the enums don't contain all actual values either (like the names for increment operators . bypassing this was just simpler (esp. as the underlying type only uses the name anyways).

CyrusNajmabadi commented:
@333fred ptal

CyrusNajmabadi commented:
@dotnet/roslyn-compiler @dotnet/roslyn-ide ptal.  thanks!

Youssef1313 commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3821,57 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            Debug.Assert(csharpLeftType is not null);
+            Debug.Assert(csharpRightType is not null);
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None)
+                throw new ArgumentException($"Illegal operator name '{name}'", nameof(name));
+
+            var binaryOperatorName = OperatorFacts.BinaryOperatorNameFromSyntaxKindIfAny(syntaxKind, isChecked);
+            if (binaryOperatorName != name)
+                throw new ArgumentException($"'{name}' was not a valid binary operator name along with isChecked={isChecked}", nameof(name));
```
Very neat!
This is so much better than my suggestion :)

CyrusNajmabadi replied:
thanks :)

Youssef1313 commented on src/Compilers/VisualBasic/Test/Semantic/Compilation/CompilationAPITests.vb:
```diff
@@ -1734,6 +1735,118 @@ BC2014: the value '_' is invalid for option 'RootNamespace'
+        <InlineData(WellKnownMemberNames.UnaryPlusOperatorName, "Public Shared Operator +(value As Integer) As Integer")>
+        <InlineData(WellKnownMemberNames.UnaryNegationOperatorName, "Public Shared Operator -(value As Integer) As Integer")>
+        <InlineData(WellKnownMemberNames.OnesComplementOperatorName, "Public Shared Operator Not(value As Integer) As Integer")>
+        <InlineData(WellKnownMemberNames.ImplicitConversionName, "Public Shared Operator CType(value As Integer) As Integer")>
+        <InlineData(WellKnownMemberNames.ExplicitConversionName, "Public Shared Operator CType(value As Integer) As Integer")>
```
I don't see C# tests for these. I also think they won't work since `UnaryOperatorNameFromSyntaxKindIfAny` doesn't seem to handle them. Not sure if this is by design?

CyrusNajmabadi replied:
c# doesn't support these afaict.  i could add tests, but just that they fail.

Youssef1313 replied:
> c# doesn't support these afaict

I thought these are:

```csharp
public class C
{
    public static implicit operator int(C c) => 0;
}
```

So I don't quite understand why they shouldn't work in C# but work in VB.
I understand that these are different languages and it's not always a one-to-one mapping, but I don't see the reason for this specific case.

CyrusNajmabadi replied:
afaict, tehse are not the same in the lnaguages, and c# doesn't represent things in that fashion.  You can see how it does it with how c# generates these SynthesizedIntrinsic operators currently.

Youssef1313 commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2909,6 +2909,44 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                returnType As ITypeSymbol,
+                leftType As ITypeSymbol,
+                rightType As ITypeSymbol,
+                isChecked As Boolean) As IMethodSymbol
```
Should we throw if this is true? there are no checked operators in VB, I think?

CyrusNajmabadi replied:
VB def has checked operators.  And it passes true here in several cases.  for example:

```vb
' for unary
Checked AndAlso operandType.IsIntegralType() AndAlso
op = UnaryOperatorKind.Minus
```

```vb
' for binary
Checked AndAlso leftType.IsIntegralType() AndAlso
(op = BinaryOperatorKind.Multiply OrElse
op = BinaryOperatorKind.Add OrElse
op = BinaryOperatorKind.Subtract OrElse
op = BinaryOperatorKind.IntegerDivide)
```

Note: it does not used the 'checked' name in this case (as that's about calling these new operators).

333fred commented on src/Compilers/Core/Portable/Symbols/MethodKind.cs:
```diff
@@ -15,6 +15,7 @@ public enum MethodKind
         /// An anonymous method or lambda expression
         /// </summary>
         AnonymousFunction = 0,
+
```
Consider reverting changes in this file.

333fred replied:
This does not appear to be addressed. The lack of space was intentional, I believe, to convey the relationship between the two kinds.

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3821,57 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            if (binaryOperatorName != name)
+                throw new ArgumentException($"'{name}' was not a valid binary operator name along with isChecked={isChecked}", nameof(name));
+
+            return new SynthesizedIntrinsicOperatorSymbol(csharpLeftType, name, csharpRightType, csharpReturnType, isChecked).GetPublicSymbol();
```
I believe this code will cause us to generate operators that do not exist. For example, using `Nullable<T>` where it should be just `T`. We have a decent amount of code in the semantic model already for synthesizing intristic operators from `BoundNode`s, it feels to me like we need to extract that into a helper that can be called from here as well.

CyrusNajmabadi replied:
>  it feels to me like we need to extract that into a helper that can be called from here as well.

Def would like your thoughts on how best to do that.  My understanding though from teh API notes was that it seemed like this was ok with people and that additional checks could be optional in the future.  :)

333fred replied:
I definitely didn't get the impression from the meeting (and it wasn't my intent to convey in the notes) that validation would optionally be added in the future. Only that we should look into it as we implemented the API. As to how to approach it: we should look at the existing places in the semantic model that create these symbols and refactor them into two components: components that extract the necessary types from the underlying bound nodes, and then the component that takes those types, does the appropriate transform (such as stripping `Nullable` where needed), and returns the resultant operator. Component 1 lives in the semantic model, 2 lives in the compilation and is called from this new api as well.

CyrusNajmabadi replied:
Cool, i'll look into that!

CyrusNajmabadi replied:
Talked to fred offline about this.  @AlekseyTs we'd like to know if you have some thoughts about this API.  For context, please see https://github.com/dotnet/roslyn/issues/63186

The idea here is to allow clients of the Compilation API to be able to generate 'built in' operators programatically, just like we can do for thigns like `Compilation.CreateArrayTypeSymbol` or `Compilation.CreateTupleType...` etc. etc.

These APIs have traditionally been used by the IDE so we can break symbols down to constituent parts, then reconstitute them later.  Over the years there have been some gaps here, which generally leads to certain types of symbols not working properly in some features.

One of these symbol types is the "built-in operator" IMethodSymbols.  There was no way to get back to an instance of this (without something really ugly like generating source-code and trying to bind that) :(.  We approved an API here, but have some general questions around how smart/dumb the API should be.  Currently, the API is very dumb.  It effectively doesn't do much validation (beyond name/null checks), and really just passes the inputs as-is down to the underlying SynthesizedIntrinsicOperator instance.   This works, and is totally fine, but does allow for generation of non-existent builtin operators (for example, you could generate `string DateTime.operator + (DateTime b, int? i)`).

There are two open questions here:

1. should we attempt to eb *very* strict, and only allow generation of built-in operators the language actually has?
2. when taking in inputs, should we transform them a-la: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs#L3770.  For example, removing nullability from types.

*personally*, i think we should not do anything but basic validation, and we should not transform.

The reason to not do anything but basic validation is to follow what we see elsewhere in the other creation APIs we have.  For example, while the language does not allow an array of ref-struct types, we are still allowed to create an IArrayTypeSymbol representing that.  Similarly, our other apis do not tend to transform inputs.  THey simply check that htey are respectively C# or VB symbols and pass them along.  

However, i am curious how you guys feel as general owners here with a lot of experience around the other creation APIs exposed off of the Compilation object.

Thanks!

CyrusNajmabadi replied:
hi @AlekseyTs Would appreciate your thoughts here. Thank you!

AlekseyTs replied:
> Would appreciate your thoughts here.

I think there are the following options:
1. No support for creating symbols like that
2. Support creation
  a. No validation - feels unnecessary relaxed.
  b. Full validation is probably ideal. But I think it will be hard to implement, because it looks like the existing code and logic is not structured in a way that allows easy reuse for validation purposes. For example, operator overload resolution looks for a suitable intrinsic operator for two expressions (not types), and the one it finds might not match the operand types exactly. Plus enum operators, UTF8 concatenation, etc.
  c. Limited validation - perhaps we can come up with not extremely complicated validation heuristic that will detect obvious errors like: use of ```Nullable<T>```, different parameter types for intrinsic types, arbitrary types (not enums, delegates, intrinsic, etc) .

CyrusNajmabadi replied:
> Limited validation - perhaps we can come up with not extremely complicated validation heuristic that will detect obvious errors like: use of Nullable<T>, different parameter types for intrinsic types, arbitrary types (not enums, delegates, intrinsic, etc) .

Let me look into this to see what i can do.  Thanks!

CyrusNajmabadi replied:
I think you do support synthesizing ops for enums btw:

![image](https://user-images.githubusercontent.com/4564579/186268800-03ad3ba8-2160-4008-b9b1-ec8e4470f195.png)

Or did you mean something else? :)

AlekseyTs replied:
> I think you do support synthesizing ops for enums btw

I think I mentioned them: "Plus **enum** operators, UTF8 concatenation, etc."

CyrusNajmabadi replied:
yeah, i read 'not enums' as 'disallow enums'.  As opposed to "don't allow aribtrary types, but *do* allow enums" :)

CyrusNajmabadi replied:
interesting thing i've found.  'dynamic' really gums up the works.  Compiler will happily synthesize intrinsic operators between it and virtually any other type i've checked.  i either have to special case dynamic (and basically then have no checks) or allow everything.

333fred reviewed (commented):
Done review pass (commit 24).

JoeRobich commented:
/azp run roslyn-integration-CI

CyrusNajmabadi commented:
ok.  i've added checks after going through and seeing what sort of checks VB and C# have in the operator space.  i also updated existing tests that look for MethodKind.BuiltInOperator and validated that all symbols we're currently testing can also go through this new API successfully.  @AlekseyTs @333fred ptal.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/BinaryOperators.vb:
```diff
@@ -1211,6 +1211,10 @@ End Class
 
+            Dim synthesizedMethod = compilation.CreateBuiltinOperator(
+                symbol1.Name, symbol1.ReturnType, symbol1.Parameters(0).Type, symbol1.Parameters(1).Type, symbol1.IsCheckedBuiltin)
+            Assert.Equal(synthesizedMethod, symbol1)
```
hooks to validate existing testcases where we create builtin-operators can properly be created with this new api.

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -7620,6 +7620,10 @@ ExpressionSyntax node4
 
+            var synthesizedMethod = compilation.CreateBuiltinOperator(
+                symbol1.Name, symbol1.ReturnType, symbol1.Parameters[0].Type, symbol1.IsCheckedBuiltin);
+            Assert.Equal(synthesizedMethod, symbol1);
```
hooks to validate existing testcases where we create builtin-operators can properly be created with this new api.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+            void CheckSupportedBinaryOperatorType(ITypeSymbol type, string paramName)
+            {
+                // Enums have operators automatically synthesized for them.
+                if (type.TypeKind == TypeKind.Enum)
```
>type.TypeKind == TypeKind.Enum [](http://example.com/codeflow?start=20&length=30)

Perhaps should let error types through. TypeKind won't be specific for them

CyrusNajmabadi replied:
Ah.  Good case.  Will do that and add tests.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+                if (type.OriginalDefinition.Equals(this.CommonGetWellKnownType(WellKnownType.System_ReadOnlySpan_T)?.GetISymbol(), SymbolEqualityComparer.ConsiderEverything) &&
+                    ((INamedTypeSymbol)type).TypeArguments[0].SpecialType == SpecialType.System_Byte)
+                {
+                    return;
```
>return; [](http://example.com/codeflow?start=20&length=7)

Is it possible to make this path C# only?

CyrusNajmabadi replied:
it is.  i was on the fence about going to that level or not.  but if you'd prefer that, i'm ok with that.

CyrusNajmabadi replied:
done.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+        /// <summary>
+        /// Creates an <see cref="IMethodSymbol"/> whose <see cref="IMethodSymbol.MethodKind"/> is <see
+        /// cref="MethodKind.BuiltinOperator"/> for a unary operator. Built-in operators are commonly created for
+        /// symbols like <c>bool int.operator==(int v1, int v2)</c> which the language implicitly supports, even if such
```
>bool int.operator==(int v1, int v2) [](http://example.com/codeflow?start=28&length=35)

Consider using a unary operator as an example here.

CyrusNajmabadi replied:
makes sense :)

CyrusNajmabadi replied:
Done.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+        /// <param name="name">The unary operator name.  Should be one of the names from <see cref="WellKnownMemberNames"/>.</param>
+        /// <param name="returnType">The return type of the unary operator.</param>
+        /// <param name="valueType">The type the operator applies to.</param>
+        /// <param name="isChecked">Whether or not this is a unary binary operator..</param>
```
>unary binary operator.. [](http://example.com/codeflow?start=61&length=23)

This looks like a typo, including double dot at the end.

CyrusNajmabadi replied:
:D

CyrusNajmabadi replied:
Fixed.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+
+            static void CheckSupportedUnaryOperatorType(ITypeSymbol type, string paramName)
+            {
+                // Delegates have operators automatically synthesized for them.
```
>Delegates [](http://example.com/codeflow?start=19&length=9)

The comment doesn't look related to the code

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+                CheckSupportedBinaryOperatorType(rightType, nameof(rightType));
+            }
+
+            return CommonCreateBuiltinOperator(name, returnType, leftType, rightType, isChecked);
```
>CommonCreateBuiltinOperator [](http://example.com/codeflow?start=19&length=27)

How hard would it be to filter operator names based on type involved? For example, logical and is not supported for integers, etc. Also some operators for special types do not allow type mix. For example, there is no operator ```int+byte```. Though rules could be language specific

CyrusNajmabadi replied:
> How hard would it be to filter operator names based on type involved? 

My main concerns there would be:

1. having to duplicate all the logic already in the compiler elsewhere in some of yoru tables (though if you can help by giving a suggestion on how to potentially just use those tables to validate things, that would make this more palatable).
2. potentially having this logic start to diverge from the actual compiler logic (Again, if we can reuse the actual logic... this gets palatable).

A good example of this might be the utf8 string work.  Imagine if this helper were written prior to that.  Do we think we would have effectively fixed this up when that was added, or we could have missed it?  I'd prefer to be a little more lenient than restrictive.  However, i don't feel strongly and would be willing to beef thigns up here (though i could use some help with that).

CyrusNajmabadi replied:
Note: this is the allure of the "allow everything" approach.  Sure, it allows for completely bogus methods to be created.  But that doesn't particularly bother me.  Then tehre is no maintenance parity concern for future changes.  I can see the desire to not allow wonky things though, but i'm wary about how complex it will be to support that.  If there's some nice way to do it though, i'm def supportive.

AlekseyTs replied:
For C#, compiler tables are more overload resolution tables. I.e., given types, give me an operator, if any. Not so much a list of supported operators. However, pretty much for all operators both parameter types should be the same (shift operators probably the only exception), and throwing this type to the overload resolution table should yield back the same type. Here are the methods that do table driven overload resolution:  [https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/BinaryOperatorEasyOut.cs,8bb194cd6f262730](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/BinaryOperatorEasyOut.cs,8bb194cd6f262730), [https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/UnaryOperatorEasyOut.cs,1dd7a6290245c3c0](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/UnaryOperatorEasyOut.cs,1dd7a6290245c3c0). It looks like it should be possible to use them as is, or create clones that operate with different input types.

AlekseyTs replied:
It looks like VB goes the same way, perhaps with additional exceptions. Functions to look at for overload resolution: [https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,333d336f7b1c2a60](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,333d336f7b1c2a60), [https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,f6c0997d84c5b913](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,f6c0997d84c5b913).

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+        /// <param name="isChecked">Whether or not this is a checked binary operator.  For example <c>int int.operator
+        /// checked +(int v1, int v2)</c>.</param>
+        /// <returns></returns>
+        public IMethodSymbol CreateBuiltinOperator(string name, ITypeSymbol returnType, ITypeSymbol leftType, ITypeSymbol rightType, bool isChecked)
```
>isChecked [](http://example.com/codeflow?start=138&length=9)

This parameter feels redundant since it is now part of the name. Or am my missing something?

CyrusNajmabadi replied:
so, it seems like in VB this parameter is provided, indepedent of the name (unless i misunderstood the code).  For example:

https://github.com/dotnet/roslyn/blob/8c927833f64f13cc3ff2650e6cf64b0144c94c3d/src/Compilers/VisualBasic/Portable/BoundTree/BoundBinaryOperator.vb#L45-L55

Here the 'checked' flag is determined with some boolean checks of the types.  But the name is chosen independently.  The name-choosing code does this:

https://github.com/dotnet/roslyn/blob/8c927833f64f13cc3ff2650e6cf64b0144c94c3d/src/Compilers/VisualBasic/Portable/Semantics/Operators.vb#L217-L267

Which does not return the "checked" names at all.

CyrusNajmabadi replied:
So i was trying to fit with that aspect for how VB/C# do things.  If we think that can be adjusted, i'm ok with it.  For example, we could have hte name make this determination, but hten in VB map the "checked" name to the "unchecked" form when actually making the method.

However this will make *using* this API much harder.  For example, if i have a VB checked built-in operator, it will give the name back as WellKnownMemberNames.AdditionOperatorName.  But if then need to *call* this api i'll have to pass along CheckedAdditionalOperatorName.  

I reallllllly wanted consumption and production of these symbols to be really easy.  The IDE is already being passed these symbols, we'd like to just grab the data we need off of them and then use that data to construct the symbols without having to worry about things potentially blowing up.

AlekseyTs replied:
> it seems like in VB this parameter is provided, indepedent of the name

Perhaps we should change that, and "teach" VB to take advantage of new names that are now part of the official set

CyrusNajmabadi replied:
will do.  Based on offline conversation, i'll make a dedicated PR for this that we can review first.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -7620,6 +7620,10 @@ ExpressionSyntax node4
             Assert.Equal(MethodKind.BuiltinOperator, symbol1.MethodKind);
             Assert.True(symbol1.IsImplicitlyDeclared);
 
+            var synthesizedMethod = compilation.CreateBuiltinOperator(
```
>CreateBuiltinOperator [](http://example.com/codeflow?start=48&length=21)

I think it makes sense to add similar validation to Microsoft.CodeAnalysis.Test.Utilities.TestOperationVisitor.VisitUnaryOperator. We have a test leg that pushes virtually every test scenario that we have in compiler's tests through that validator. So, we should be able to catch any form of unary operator for which we have a test today, or will have a test in the future.

AlekseyTs replied:
Note, that IOperation nodes probably don't return symbols for intrinsic operators, so they will have to be requested from semantic model based on syntax.

CyrusNajmabadi replied:
done.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -8396,6 +8400,10 @@ ExpressionSyntax node8
             Assert.Equal(MethodKind.BuiltinOperator, symbol1.MethodKind);
             Assert.True(symbol1.IsImplicitlyDeclared);
 
+            var synthesizedMethod = compilation.CreateBuiltinOperator(
```
>CreateBuiltinOperator [](http://example.com/codeflow?start=48&length=21)

Similarly, we should consider adding similar validation to Microsoft.CodeAnalysis.Test.Utilities.TestOperationVisitor.VisitBinaryOperator

CyrusNajmabadi replied:
done.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,153 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+                //
+                // For error types, we don't bother checking.  The type was already in error, so fine to create an error
+                // builtin wrapping it.
+                if (type.TypeKind is TypeKind.Delegate or TypeKind.Enum or TypeKind.Error or TypeKind.Pointer)
```
>or TypeKind.Error [](http://example.com/codeflow?start=72&length=17)

An error type with a known SpecialType should still pass the SpecialType filter

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1488,6 +1489,153 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+                //
+                // For error types, we don't bother checking.  The type was already in error, so fine to create an error
+                // builtin wrapping it.
+                if (type.TypeKind is TypeKind.Dynamic or TypeKind.Enum or TypeKind.Error or TypeKind.Pointer)
```
>TypeKind.Error [](http://example.com/codeflow?start=74&length=14)

An error type with a known SpecialType should still pass the SpecialType filter

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3821,64 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        protected override bool IsLanguageSpecificSupportedBuiltinOperatorType(ITypeSymbol type)
+        {
+            // ReadOnlySpan<byte> supports a builtin `+` operator.
+            return type.OriginalDefinition.Equals(this.CommonGetWellKnownType(WellKnownType.System_ReadOnlySpan_T)?.GetISymbol(), SymbolEqualityComparer.ConsiderEverything) &&
```
>CommonGetWellKnownType [](http://example.com/codeflow?start=55&length=22)

Rather than going through the common layer, consider dropping to the C# specific layer instead, less translations will be involved.

AlekseyTs commented:
Done with review pass (commit 43)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -2188,44 +2188,6 @@ internal static SpecialType GetEnumPromotedType(SpecialType underlyingType)
 
-        private static BinaryOperatorKind SyntaxKindToBinaryOperatorKind(SyntaxKind kind)
-        {
-            switch (kind)
```
moved to OperatorFacts to help out.

AlekseyTs replied:
> moved to OperatorFacts to help out.

Please leave the helper here in it original form, just make it accessible to other components

dibarbet commented:
/azp run

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,276 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    }
+                }
+
+                // bool operator ==(object, object) is legal.
```
the following was me reimplementing the logic from BinaryOperatorOverloadResolution_NoEasyOut.  I couldn't find a way to do it in a cleaner fashion given taht that logic depends on BoundExpressions, not types.

CyrusNajmabadi commented:
@AlekseyTs @333fred this is ready for review.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            string name,
+            ITypeSymbol returnType,
+            ITypeSymbol leftType,
+            ITypeSymbol rightType)
```
base type does the null checks.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    return;
+                }
+
+                // T* operator+(int i, T&)
```
```suggestion
                // T* operator+(int i, T*)
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Operators/OperatorFacts.cs:
```diff
@@ -210,5 +210,44 @@ public static string BinaryOperatorNameFromOperatorKind(BinaryOperatorKind kind,
             }
         }
+
+        public static BinaryOperatorKind SyntaxKindToBinaryOperatorKind(SyntaxKind kind)
```
>public static BinaryOperatorKind SyntaxKindToBinaryOperatorKind(SyntaxKind kind) [](http://example.com/codeflow?start=8&length=80)

Please move helpers to their original place in their original syntactic form (no behavior changes - no code changes), simply adjust their accessibility as appropriate.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            void checkOperatorKinds()
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (isDynamicOrError(csharpReturnType) ||
```
>isDynamicOrError(csharpReturnType) [](http://example.com/codeflow?start=20&length=34)

I think error types with known special type should go through additional validation. Effectively we know what they are.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            static bool isDynamicOrError(TypeSymbol type)
+                => type.TypeKind is TypeKind.Dynamic or TypeKind.Error;
+
+            void checkOperatorKinds()
```
>checkOperatorKinds [](http://example.com/codeflow?start=17&length=18)

"validateSignature"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                if (easyOutBinaryKind != BinaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutBinaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
```
>TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=24&length=93)

I think we can simply compare SpecialTypes instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                // Actual delegates have several operators that can be used on them.
+                if (csharpLeftType.TypeKind is TypeKind.Delegate &&
+                    TypeSymbol.Equals(csharpLeftType, csharpRightType, TypeCompareKind.ConsiderEverything))
```
>TypeSymbol.Equals(csharpLeftType, csharpRightType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=20&length=86)

Compare SpecialType?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    if (binaryKind is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual &&
+                        csharpReturnType.SpecialType == SpecialType.System_Boolean)
+                    {
+                        return;
```
>return [](http://example.com/codeflow?start=24&length=6)

Is this code path reachable? It looks like comparison of delegates is handled above. Together with object comparison.

CyrusNajmabadi replied:
one is literally System.Delegate, teh other are acutal delegates (like Action).

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    // SomeDelegate operator +(SomeDelegate, SomeDelegate) is legal.
+                    // SomeDelegate operator -(SomeDelegate, SomeDelegate) is legal.
+                    if (binaryKind is BinaryOperatorKind.Addition or BinaryOperatorKind.Subtraction &&
+                        TypeSymbol.Equals(csharpLeftType, csharpReturnType, TypeCompareKind.ConsiderEverything))
```
>TypeSymbol.Equals(csharpLeftType, csharpReturnType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=24&length=87)

Compare SpecialType?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    if (binaryKind is BinaryOperatorKind.Addition or BinaryOperatorKind.Subtraction)
+                    {
+                        if (csharpLeftType.IsEnumType() && csharpRightType.IsIntegralType() && TypeSymbol.Equals(csharpLeftType, csharpReturnType, TypeCompareKind.ConsiderEverything))
+                            return;
```
>return; [](http://example.com/codeflow?start=28&length=7)

A block is required around ```if``` statement body.

CyrusNajmabadi replied:
Nope :)  compiler and IDE moved to .net rules, which don't require curlies on it-statements.  IDE used to actually require this, but we stopped at request of compiler team which wanted to follow .net rules :)

AlekseyTs replied:
Please put them anyway. Our rule was, and I am not aware that we changed it - multiline if must use blocks

AlekseyTs replied:
Having blocks doesn't violate the style. Right?

CyrusNajmabadi replied:
I'll add back in.  But this relaxation was made at teh request of compiler team :)   IDE always had this as a hard requirement, but we had to relax it because of this desire to match .net formatting.

CyrusNajmabadi replied:
I'm just saying, we had such a rule, but it was actually removed at the request *of hte compiler team*.  Specifically just so this sort of block could be written.  It's strange to have had that happen, and now get requests to be more restrictive with the formatting, as taht was the position we were pushing for back then anyways, and which we were overruled on :)  

One of the takeaways was that when it came to formatting we would simply follow the .net rules and have tooling check/correct when in violation.  That was to help prevent these sorts of discussions from arising and it's why we settled on .net rules as well.  It feels like that has now been rolled back as we're instituting stricter rules, and we *are* spending the time in PRs discussing this, which was the thing that the team wanted to stop happening :)

AlekseyTs replied:
And I am just saying that we had the rule and, even though I am a member of the compiler team, I am not aware of a change made around this rule or any request that the team made about that. I am not saying that no one asked for the change on behalf of the compiler team.

CyrusNajmabadi replied:
@jaredpar for the history here :)

AlekseyTs replied:
If you are fine with adding blocks, I am not sure what makes this thread going.

CyrusNajmabadi replied:
ideally, to prevent hte discussion in future PRs.  I'm fine with adding the blocks to lubricate this PR.  Generally speaking though, i will want to continue following any legal .net style, which includes not having blocks when not required :)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    // SomeEnum operator-(int, int)
+                    if (binaryKind is BinaryOperatorKind.Addition or BinaryOperatorKind.Subtraction)
+                    {
+                        if (csharpLeftType.IsEnumType() && csharpRightType.IsIntegralType() && TypeSymbol.Equals(csharpLeftType, csharpReturnType, TypeCompareKind.ConsiderEverything))
```
>IsIntegralType [](http://example.com/codeflow?start=75&length=14)

The check doesn't match the comment. This isn't a check for ```int```, I believe.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    // SomeEnum operator+(SomeEnum, int)
+                    // SomeEnum operator+(int, SomeEnum)
+                    // SomeEnum operator-(SomeEnum, int)
+                    // SomeEnum operator-(int, int)
```
>SomeEnum operator-(int, int) [](http://example.com/codeflow?start=23&length=28)

This looks suspicious. Typo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                        if (csharpLeftType.IsEnumType() && csharpRightType.IsIntegralType() && TypeSymbol.Equals(csharpLeftType, csharpReturnType, TypeCompareKind.ConsiderEverything))
+                            return;
+
+                        if (csharpRightType.IsEnumType() && csharpLeftType.IsIntegralType() && TypeSymbol.Equals(csharpRightType, csharpReturnType, TypeCompareKind.ConsiderEverything))
```
>IsIntegralType [](http://example.com/codeflow?start=75&length=14)

The check doesn't match the comment. This isn't a check for int, I believe.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                    // int operator-(SomeEnum, SomeEnum)
+                    if (binaryKind is BinaryOperatorKind.Subtraction &&
+                        csharpReturnType.IsIntegralType() &&
```
>IsIntegralType [](http://example.com/codeflow?start=41&length=14)

The check doesn't match the comment. This isn't a check for int, I believe.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                // T* operator+(T*, int i)
+                if (binaryKind is BinaryOperatorKind.Addition &&
+                    csharpLeftType.IsPointerType() &&
+                    csharpRightType.IsIntegralType() &&
```
>IsIntegralType [](http://example.com/codeflow?start=36&length=14)

The check doesn't match the comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                // T* operator+(int i, T*)
+                if (binaryKind is BinaryOperatorKind.Addition &&
+                    csharpRightType.IsPointerType() &&
+                    csharpLeftType.IsIntegralType() &&
```
>IsIntegralType [](http://example.com/codeflow?start=35&length=14)

The check doesn't match the comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                // T* operator-(T*, int i)
+                if (binaryKind is BinaryOperatorKind.Subtraction &&
+                    csharpLeftType.IsPointerType() &&
+                    csharpRightType.IsIntegralType() &&
```
>IsIntegralType [](http://example.com/codeflow?start=36&length=14)

The check doesn't match the comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                // ROS<byte> operator+(ROS<byte>, ROS<byte>). Legal because of utf8 strings.
+                if (binaryKind is BinaryOperatorKind.Addition &&
+                    this.IsReadOnlySpanType(csharpReturnType) &&
```
>this.IsReadOnlySpanType(csharpReturnType) [](http://example.com/codeflow?start=20&length=41)

The check doesn't look strict enough. There is no check for byte.

CyrusNajmabadi replied:
whoops!

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            Debug.Assert(csharpOperandType is not null);
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None || name is WellKnownMemberNames.TrueOperatorName or WellKnownMemberNames.FalseOperatorName)
```
>name is WellKnownMemberNames.TrueOperatorName or WellKnownMemberNames.FalseOperatorName [](http://example.com/codeflow?start=49&length=87)

It is not obvious why these are excluded. Is the reason that compiler never creates symbols for them? Consider adding a clarifying comment.

AlekseyTs replied:
Note that the exception is likely to confuse the consumer, should it refer specifically to "built-in operator"

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None)
+                throw new ArgumentException($"Illegal operator name '{name}'", nameof(name));
```
>ArgumentException [](http://example.com/codeflow?start=26&length=17)

Note that majority of messages for exceptions thrown in this file are localizable, i.e. resource based.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            static bool isDynamicOrError(TypeSymbol type)
+                => type.TypeKind is TypeKind.Dynamic or TypeKind.Error;
+
+            void checkOperatorKinds()
```
>checkOperatorKinds [](http://example.com/codeflow?start=17&length=18)

"validateSignature"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (isDynamicOrError(csharpReturnType) ||
+                    isDynamicOrError(csharpOperandType))
```
>isDynamicOrError(csharpOperandType) [](http://example.com/codeflow?start=20&length=35)

I think Error types with known SpecialType,should go through stricter rules, we know what they are.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                if (easyOutUnaryKind != UnaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutUnaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
```
>TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=24&length=93)

Switch to special type comparison?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                if (easyOutUnaryKind != UnaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutUnaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
```
>TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=24&length=93)

Does this prevent nullable types? Do we have a test for a scenario like that?

CyrusNajmabadi replied:
yes. CreateBuiltinBinaryOperator_NotSupported checks for nullable types in any position.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                // Use fast-path check to see if this types are ok.
+                var binaryKind = OperatorFacts.SyntaxKindToBinaryOperatorKind(SyntaxFacts.GetBinaryExpression(syntaxKind));
+                var easyOutBinaryKind = OverloadResolution.BinopEasyOut.OpKind(binaryKind, csharpLeftType, csharpRightType);
```
>var easyOutBinaryKind = OverloadResolution.BinopEasyOut.OpKind(binaryKind, csharpLeftType, csharpRightType); [](http://example.com/codeflow?start=16&length=108)

It feels like this should be done only when both operands have SpecialType

CyrusNajmabadi replied:
i considered that, but it felt redudant given OpKind's behavior.  is there a case we're concerned about here?

AlekseyTs replied:
>  is there a case we're concerned about here?

Extra unnecessary work and lack of clarity what inputs can pass this check.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                // Use fast-path check to see if this types are ok.
+                var unaryKind = OperatorFacts.SyntaxKindToUnaryOperatorKind(SyntaxFacts.GetPrefixUnaryExpression(syntaxKind));
+                var easyOutUnaryKind = OverloadResolution.UnopEasyOut.OpKind(unaryKind, csharpOperandType);
```
>var easyOutUnaryKind = OverloadResolution.UnopEasyOut.OpKind(unaryKind, csharpOperandType); [](http://example.com/codeflow?start=16&length=91)

It feels like this should be done only when operand has SpecialType

AlekseyTs commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -735,6 +735,10 @@ public override void VisitUnaryOperator(IUnaryOperation operation)
+
+            // Directly get the symbol for this operator from the semantic model.  This allows us to exercise
+            // potentially creating synthesized intrinsic operators.
+            operation.SemanticModel?.GetSymbolInfo(operation.Syntax);
```
>operation.SemanticModel?.GetSymbolInfo(operation.Syntax); [](http://example.com/codeflow?start=12&length=57)

I though the intent was to not just get the symbol, but also recreate it if it is built-in.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -768,6 +772,10 @@ public override void VisitBinaryOperator(IBinaryOperation operation)
+
+            // Directly get the symbol for this operator from the semantic model.  This allows us to exercise
+            // potentially creating synthesized intrinsic operators.
+            operation.SemanticModel?.GetSymbolInfo(operation.Syntax);
```
>operation.SemanticModel?.GetSymbolInfo(operation.Syntax); [](http://example.com/codeflow?start=12&length=57)

I though the intent was to not just get the symbol, but also recreate it if it is built-in.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2909,6 +2910,137 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            ' Quick table access to determine if these types are legal.
+            Dim resolved = OverloadResolution.ResolveNotLiftedIntrinsicBinaryOperator(opInfo.BinaryOperatorKind, leftType.SpecialType, rightType.SpecialType)
+            If resolved <> SpecialType.None Then
+                ' Quick access table strangely maps `string Like string` to the `string` return type. remap it to 'bool'
```
>Quick access table strangely maps `string Like string` to the `string` return type. remap it to 'bool' [](http://example.com/codeflow?start=18&length=102)

I think it actually maps it to the operand type, which for most operators is equal to return type. However, there are outliers.

CyrusNajmabadi replied:
ok.  i misunderstood teh table then.  have adjusted the callsite accordingly.   will see if tests reveal any other things.

AlekseyTs replied:
>  have adjusted the callsite accordingly.

Not sure if any adjustments were necessary apart from comments, perhaps

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2909,6 +2910,137 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            Dim vbReturnType = returnType.EnsureVbSymbolOrNothing(Of TypeSymbol)(NameOf(returnType))
+            Dim vbOperandType = returnType.EnsureVbSymbolOrNothing(Of NamedTypeSymbol)(NameOf(operandType))
+
+            Dim nameToCheck = If(SynthesizedIntrinsicOperatorSymbol.IsCheckedUnaryOperator(name), WellKnownMemberNames.UnaryNegationOperatorName, name)
```
>SynthesizedIntrinsicOperatorSymbol.IsCheckedUnaryOperator(name) [](http://example.com/codeflow?start=33&length=63)

It is probably better to check for ```CheckedUnaryNegationOperatorName``` explicitly here rather than delegating to some other helper with general enough name to question the logic that follows.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2909,6 +2910,137 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                End If
+            End If
+
+            Throw New ArgumentException($"Unsupported built-in binary operator: {returnType.ToDisplayString()} operator {name}({leftType.ToDisplayString()}, {rightType.ToDisplayString()})")
```
>Throw [](http://example.com/codeflow?start=12&length=5)

Are we going to get here for error types?

CyrusNajmabadi replied:
Yes.  This will now throw.  We will only defer to SpecialType if present.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2909,6 +2910,137 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                opInfo As OverloadResolution.OperatorInfo)
+            ' 
+            If returnType.IsErrorType() OrElse operandType.IsErrorType() Then
+                Return
```
>Return [](http://example.com/codeflow?start=16&length=6)

I think that operadndType with known SpecialType should still pass the tests below. We know what that type is.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SynthesizedSymbols/SynthesizedIntrinsicOperatorSymbol.vb:
```diff
@@ -29,19 +29,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             _parameters = (New ParameterSymbol() {New SynthesizedOperatorParameterSymbol(Me, container, 0, "value")}).AsImmutableOrNull()
         End Sub
 
+        Public Shared Function IsCheckedUnaryOperator(name As String) As Boolean
```
>IsCheckedUnaryOperator [](http://example.com/codeflow?start=31&length=22)

Given a comment at the call site, this refactoring is probably unnecessary.

AlekseyTs commented:
Done with review pass (commit 99)

CyrusNajmabadi commented:
@AlekseyTs All feedback responded to, i believe :)

AlekseyTs commented:
@CyrusNajmabadi It looks like there are legitimate test failures

AlekseyTs commented:
@CyrusNajmabadi It looks like IOperation leg caught some failures.

AlekseyTs commented:
@CyrusNajmabadi It looks like there are still failures

CyrusNajmabadi commented:
@AlekseyTs I'm making changes and sending them through CI to discover the failures.  Running locally takes too long.

CyrusNajmabadi commented:
@AlekseyTs @333fred this is ready for review. Thanks!

CyrusNajmabadi commented:
@AlekseyTs @333fred this is ready for eyes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,300 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    return;
+                }
+
+                throw new ArgumentException(string.Format(CodeAnalysisResources.Unsupported_built_in_operator_0, $"{csharpReturnType.ToDisplayString()} operator {name}({csharpLeftType.ToDisplayString()}, {csharpRightType.ToDisplayString()})"));
```
>throw new ArgumentException [](http://example.com/codeflow?start=16&length=27)

Do we want some scenarios involving error types to pass through here? 
We have checks for TypeKind with the goal to identify delegates and enums. If type is an error type (perhaps cannot be found in a new compilation), those checks going to fail. So, we might want to return if at least one operand type (intentionally excluding return type) is an error type with no SpecialType.

CyrusNajmabadi replied:
> Do we want some scenarios involving error types to pass through here?

I'm completely ok with the current system as is.  IDE has no need for the error case as of yet, and none of the compiler test cases hit that.  So i'd rather make any changes here based on a needed use case.  :)

AlekseyTs replied:
> I'm completely ok with the current system as is.  IDE has no need for the error case as of yet, and none of the compiler test cases hit that.

Sounds good. just wanted to make sure a possibility of a failure like that is understood.

CyrusNajmabadi replied:
👍

AlekseyTs commented on src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:
```diff
@@ -4,7 +4,7 @@
 #nullable disable
 // Uncomment to enable the IOperation test hook on all test runs. Do not commit this uncommented.
-//#define ROSLYN_TEST_IOPERATION
+// #define ROSLYN_TEST_IOPERATION
```
It doesn't look like there are meaningful changes in this file. Consider reverting.

333fred replied:
This does not appear to be addressed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SynthesizedSymbols/SynthesizedIntrinsicOperatorSymbol.vb:
```diff
@@ -31,12 +31,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         Public Overrides ReadOnly Property IsCheckedBuiltin As Boolean
             Get
-                Select Case Me.Name
+                Select Case Name
```
>Select Case Name [](http://example.com/codeflow?start=16&length=16)

It looks like there are no meaningful changes in this file. Consider reverting.

333fred replied:
This does not appear to be addressed.

AlekseyTs reviewed (approved):
LGTM (commit 134). Please squash commits during merging.

333fred commented on src/Compilers/Core/Portable/CodeAnalysisResources.resx:
```diff
@@ -734,4 +734,13 @@
     <value>Total generator execution time: {0} seconds.</value>
   </data>
-</root>
+  <data name="Illegal_built_in_operator_name_0" xml:space="preserve">
```
Please follow compiler conventions on resource names.

CyrusNajmabadi replied:
Fwiw, I cannot figure out what your convention is.  There seems to be absolutely no relation between the identifiers and the content. :-/

CyrusNajmabadi replied:
All three of these actually follow a convention that keeps things in sync and make the consumption side clear and trivial to verify. I would prefer to keep as is.

333fred replied:
But it's inconsistent with all the rest of our code. I would suggest `IllegalBuiltInOperatorName` for this one.

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,300 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                if (binaryKind is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual &&
+                    csharpReturnType.SpecialType is SpecialType.System_Boolean)
+                {
+                    if ((csharpLeftType.SpecialType is SpecialType.System_Object && csharpRightType.SpecialType is SpecialType.System_Object) ||
+                        (csharpLeftType.SpecialType is SpecialType.System_Delegate && csharpRightType.SpecialType is SpecialType.System_Delegate))
```
Small suggestion: could make this a bit nicer with tuple patterns:

```cs
(csharpLeftType.SpecialType, csharpRightType.SpecialType) is (SpecialType.System_Object, SpecialType.System_Object)
                                                          or (SpecialType.System_Delegate, SpecialType.System_Delegate)
```

CyrusNajmabadi replied:
nice!

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3821,6 +3820,300 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                if (csharpLeftType.IsEnumType() || csharpRightType.IsEnumType())
+                {
+                    // bool operator ==(SomeEnum, SomeEnum) is legal.
```
Nit: for some of these comments, you list all combos, and for some you list just one.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:
```diff
@@ -287,7 +287,8 @@ static class Ext
         }
 
-        [ConditionalFact(typeof(IsRelease))]
+        // Too slow to run with IOperationValidation on.
```
Really? There are no unary or binary operators in this test, what caused the time it takes to increase?

CyrusNajmabadi replied:
will try reactivating to see if it passes.

333fred commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2720,6 +2720,306 @@ private static ImmutableArray<CodeAnalysis.NullableAnnotation> GetAnonymousTypeN
+        {
+            var compilation = CreateCompilation("");
+            var fakeIntType = compilation.CreateErrorTypeSymbol(compilation.CreateErrorNamespaceSymbol(compilation.GlobalNamespace, "System"), "Int32", arity: 0);
+            Assert.Throws<ArgumentException>(() =>
```
Consider using the `Assert.Throws` variant that verifies argument name.

333fred commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2720,6 +2720,306 @@ private static ImmutableArray<CodeAnalysis.NullableAnnotation> GetAnonymousTypeN
+            var nullableIntType = compilation.GetSpecialType(SpecialType.System_Nullable_T).GetPublicSymbol().Construct(intType);
+
+            // vb binary operator name
+            Assert.Throws<ArgumentException>(() =>
```
Same for these.

333fred commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2720,6 +2720,306 @@ private static ImmutableArray<CodeAnalysis.NullableAnnotation> GetAnonymousTypeN
+        {
+            var compilation = CreateCompilation("");
+            var intType = compilation.GetSpecialType(SpecialType.System_Int32).GetPublicSymbol();
+            Assert.Throws<ArgumentNullException>(() => compilation.CreateBuiltinOperator(
```
Same for these.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/Compilation/CompilationAPITests.vb:
```diff
@@ -1734,6 +1732,207 @@ BC2014: the value '_' is invalid for option 'RootNamespace'
+            Dim nullableIntType = compilation.GetSpecialType(SpecialType.System_Nullable_T).Construct(intType)
+
+            ' c# binary operator name
+            Assert.Throws(Of ArgumentException)(Function() compilation.CreateBuiltinOperator(WellKnownMemberNames.IncrementOperatorName, intType, intType, intType))
```
Similar `Assert.Throws` comments as in C#.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/MethodKind.cs:
```diff
@@ -15,6 +15,7 @@ public enum MethodKind
         /// An anonymous method or lambda expression
         /// </summary>
         AnonymousFunction = 0,
+
```
```suggestion
```

CyrusNajmabadi commented:
@davidwengier Can you look at the IDE portion?

ryzngard commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.BuiltinOperatorSymbolKey.cs:
```diff
@@ -0,0 +1,76 @@
+                    static (contextualSymbol, i) => SafeGet(contextualSymbol.Parameters, i)?.Type,
+                    out var parameterTypesFailureReason);
+
+                if (returnTypeFailureReason != null)
```
Can we check this before reading parameterTypes?

davidwengier replied:
You have to read what you wrote, regardless of whether you could do anything with it, or the position of the stream will be wrong.

CyrusNajmabadi replied:
great question!  the answer is 'no'. You can think of the 'reading' as moving a pointer through the stream.  we have to always read to the end (even on failure).  this is because higher parts may re-read pieces (As they're resolving against different symbols).  It's definitely subtle.

ryzngard reviewed (approved):
IDE side LGTM
